src/main/java/org/nypl/findawayclientlibrary/DownloadService.java

```diff
@@ -141,19 +183,253 @@ public void onNext(Integer progress) {
+   */
+  public Integer downloadAudio(String contentId, String licenseId, Integer chapter, Integer part) {
+    if (part == null) {
+      // most of the time, we don't need to get granular.  start download at part==0, and auto-proceed from there.
+      part = CHAPTER_PART_DEFAULT;
+    }
+
+    if ((contentId == null) || (licenseId == null) || (chapter == null)) {
+      LogHelper.e(TAG, "DownloadService cannot build download request for (contentId, licenseId, chapter)", contentId, licenseId, chapter);
+      return DOWNLOAD_ERROR;
+    }
+
+    // NOTE:  DownloadType.TO_END gets book from specified chapter to the end of the book.
+    // DownloadType.TO_END_WRAP gets from specified chapter to the end of the book, then wraps around and
+    // gets all the beginning chapters, too.
+    // DownloadType.SINGLE gets just that chapter.
+    // The system does skip any chapters that are already downloaded.  So, if we need to re-download a chapter,
+    // we'd have to delete it first then call download.
+
+    LogHelper.d(TAG, "before building downloadRequest, part=" + part + ", chapter=" + chapter);
+    // NOTE:  If you start with the first chapter in your request the DownloadEngine will skip any chapters that are already downloaded.
+    DownloadRequest downloadRequest = DownloadRequest.builder().contentId(contentId).part(part).chapter(chapter).licenseId(licenseId).type(DownloadType.TO_END_WRAP).build();
+
+    try {
+      this.getDownloadEngine().download(downloadRequest);
+      LogHelper.d(TAG, "after downloadEngine.download \n\n\n");
+    } catch (Exception e) {
+      LogHelper.e(TAG, e, "Error getting download engine: " + e.getMessage());
+      return DOWNLOAD_ERROR;
+    }
+
+    return DOWNLOAD_RUNNING;
+  }
+
+
+  /**
+   * Delete all content previously downloaded for this book.
+   * TODO: untested
+   *
+   * @param contentId  book to delete
+   * @return
+   */
+  public Integer deleteDownload(String contentId) {
+    DeleteRequest deleteRequest = DeleteRequest.builder().contentId(contentId).build();
+
+    try {
+      this.getDownloadEngine().delete(deleteRequest);
+    } catch (Exception e) {
+      LogHelper.e(TAG, e, "Error deleting a download: " + e.getMessage());
+      return DOWNLOAD_ERROR;
+    }
+
+    return DELETE_REQUESTED;
+  }
+
+
+  /**
+   * Go through all download requests currently on the engine, and
+   * return those whose contentId matches the one passed in.
+   *
+   * @param contentId
+   * @return
+   */
+  public List<DownloadRequest> findDownloadRequests(String contentId) {
+    List<DownloadRequest> foundRequests = null;
+    if (contentId == null) {
+      return foundRequests;
+    }
+
+    // To get all DownloadRequests:
+    List<DownloadRequest> requests = this.getDownloadEngine().downloadRequests();
+    for (DownloadRequest request : requests) {
+      if (((DownloadRequest) request).contentId().equals(contentId)) {
+        foundRequests.add(request);
```

Comment:
`foundRequests` is always `null` here so this is always a null pointer exception. I think you need to initialize the list on line 324.

Reply:
good catch